### DISCUSSION TITLE
Generalize plugin logic

### DISCUSSION THREAD

**Discussion by DennyFiguerres** [OData/RESTier#736] [Ideas]
i like the current support for adding logic with the On<Update|Insert|Delete><Entityset>()
but what about some thing more like this as an option
OnInsert<t>(<T> item) where T impliments interface IRecord

that is not exactly the right syntax ....

the idea is that if i have adatabase with say 20 tables that all have some comon colums like date created, date changed , who updated and who created that all of the ef models could impliment an interface IRecord
if the item to be updated is compatable with IRecord then i can apply an OnUpdate function generically to all the entities that impliment that interface.

so in place of having to create 20 OnUpdateing<tablename> functions i create one that works for all of them.

i hope the idea is explained well enough here for others to see how this might work and be very usefull.

**Comment by DennyFiguerres** [comment]
well crap i see that it might already be there .... so this may be a done thing.... oh heck!

**Comment by robertmclaws** [comment]
This is 100% possible. I built a system for my company that does exactly this. It is a set of Managers that allow objects that inherit from a specific type and conform to specific interfaces to be modified correctly without the need for extra code.

We then created a manual code generator to route Restier calls through those Managers, allowing you to include additional functionality through partial methods.

The sum total of that system allows you to run the exact same business logic whether you need to interact with the database more directly (like in complex Restier Operations) or call Restier through the web.

This is how we're able to build and evolve BurnRate so quickly (https://dotnet.microsoft.com/en-us/platform/customers/burnrate)

My next goal is to turn the manual code generator into a Roslyn Source Generator so that the code just always stays updated.

I would like to productize it sometime in 2024, but that depends on a number of factors.

HTH!

**Reply by DennyFiguerres** [reply]
on the old doc site i found info about a this but i am having trouble followingit....
 IChangeSetPreparer    IChangeSetItemProcessor
reading here:
https://odata.github.io/RESTier/v0.6/#02-09-Customize-Submit

where it says:

Customized Process Logic

User can use interface IChangeSetItemProcessor to customize logic before or after submit, OnProcessingChangeSetItemAsync logic is called before submit and OnProcessedChangeSetItemAsync logic is called after submit, RESTier default logic is defined in section 2.3 plugin user logic part. Default logic can be called via defined a property with type IChangeSetItemProcessor like “private IChangeSetItemProcessor InnerProcessor {get; set;}”, and user call InnerProcessor.OnProcessingChangeSetItemAsync or OnProcessedChangeSetItemAsync to call RESTier logic, if in CustomizedProcessor, there is no such property defined or InnerProcessor is not used, then RESTier logic will not be called.

have to run to meeting but if you can help me clear this up it will be great!

**Reply by robertmclaws** [reply]
Restier makes it possible to swap out ALL of its components, but it is not for the faint of heart.

I don't believe the ChangeSetItemProcessor exists anymore. I don't remember what the equivalent is.

To prove success at whatever you're trying to do, I would highly recommend first just working within the means of the built-in convention system. Once you get that working, you can follow the requests through the pipeline and see if there isn't a better point of abstraction higher up on the stack.

The reason I say that is because we're very likely to redesign the whole system in v2 to simplify the pipeline and make it easier to swap out parts. Right now the system was designed in 2011 or something like that, so it's kind of a mess.

**Reply by DennyFiguerres** [reply]
thanks for the advise!  

and i had much the thought that you just said... i know over time i have had places where i did something and later wish i had done it another way....

for now i will see if i can do some editing tricks with notepad++ to get it to make a bunch of the code for V1 of the api set.
20 ish tables that each need a common pattern for insert, update and delete....  
so 20 * 3 functions is a bit but i can manage it (I hope)

**Comment by DennyFiguerres** [comment]
one other thing that perhaps you can point out to me:
when the api needs to have an authorized user how to pickup that principal in the api class ??
i need to set Entity.WhoChanged to something like "joe@domain.com" all over the place....

**Reply by DennyFiguerres** [reply]
is it this: string uname = ClaimsPrincipal.Current.Identity.Name; ?

**Reply by robertmclaws** [reply]
Are you on .NET Framework or .NET Core?

**Reply by DennyFiguerres** [reply]
dot net core 7 and looking to get to 8 very soon.

also by the way I am using a tool called RadZen for the Web UI in Balzor and it supports using OData.

**Reply by DennyFiguerres** [reply]
by the way i find it interesting the way that it seems like there is not much in the way of tools and examples for using OData with swagger and the rest of the microsoft systems when i know that OData is so core to much of the Microsoft Dynamics and 360 server side.

seems to always be a big gap in the tools for OData and the rest of restfull api's

**Reply by robertmclaws** [reply]
I agree! Hoping to be a part of changing that with Restier 2.0 and .NET 8 & 9.

**Reply by robertmclaws** [reply]
BTW if you're on .NET Core, you'll need to use the `.UseClaimsPrincipals()` extension in your app registration if you want to get the user in Restier, since it doesn't inject the HttpContext. That extension method will pull it out of DI for you.
